README.md
# N4B PHP Library

Simple library  to quickly implement a N4B Be-App PHP backend

[![Latest Stable Version](https://img.shields.io/packagist/v/n4b/n4b.svg)](https://packagist.org/packages/n4b/n4b)
[![Build Status](https://img.shields.io/travis/kinobi/n4b/master.svg)](https://travis-ci.org/kinobi/n4b)
[![SensioLabs Insight](https://img.shields.io/sensiolabs/i/5960a9bc-9e87-4429-a714-b2ee05c4feac.svg?maxAge=2592000)](https://insight.sensiolabs.com/projects/5960a9bc-9e87-4429-a714-b2ee05c4feac)
[![Dependencies](https://img.shields.io/versioneye/d/user/projects/57e0c530bd6fa6004e11e634.svg)](https://www.versioneye.com/user/projects/57e0c530bd6fa6004e11e634?child=summary)
[![GitHub issues](https://img.shields.io/github/issues/kinobi/n4b.svg)](https://github.com/kinobi/n4b/issues)
[![License](https://img.shields.io/packagist/l/n4b/n4b.svg)](https://packagist.org/packages/n4b/n4b)

## Why ?

N4B PHP will allow you to concentrate your development only on your business domain. Authentication, 
operations dispatching, N4B Errors, formatting of the responses are handled by the library.

### What is N4B by Be-Bound ?
[Be-Bound's N4B platform](https://n4b.io) helps you develop Android applications with the possibility to **reach the Next 4 Billion** (N4B) users that are still not connected to the Internet. 
By integrating [Be-Bound](https://www.be-bound.com)’s SDK into your Android apps, and implementing your Be-App Webhooks, **your users will stay connected to your apps even when there is no internet**.


### Learn More about N4B by Be-Bound
- [N4B Documentation](http://doc.n4b.io)
- [N4B Developer Console](https://dev.n4b.io)


## Installation
N4B PHP works with PHP 5.6.0 or newer.

It's recommended that you use [Composer](https://getcomposer.org/) to install N4B PHP.
```bash
$ composer require n4b/n4b
```
This will install N4B PHP and all required dependencies.


## Usage
Create an index.php file with the following contents:
```php
<?php

require 'vendor/autoload.php';

$n4b = new N4B\Webhook('mybeapp', 1337, 1, 'MY5UP3r53Cr37K3Y');

$n4b->add('myOperation', function($params, $transport, $userId) {
	return ['uppercaseString' => strtoupper($params['someString'])];
});

$n4b->run();
```
You may quickly test this using the built-in PHP server:
```bash
$ php -S localhost:8000
```

Then running the cURL POST request below (or use the [Postman](https://www.getpostman.com) button) will now return a well formated N4B response "{"params":["uppercaseString","THIS VERY INTERESTING TEXT WILL BE OUTPUT IN UPPERCASE"]}".

[![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/02052ba017f0c817035e)

```bash
curl --request POST \
  --url http://localhost:8000/ \
  --header 'authorization: Basic bXliZWFwcF8xMzM3Ok1ZNVVQM3I1M0NyMzdLM1k=' \
  --header 'cache-control: no-cache' \
  --header 'content-type: application/json' \
  --data '{"transport":"web","userId":"98e866b2-3d39-11e6-97dc-0cc47a77819c","moduleId":1337,"moduleName":"mybeapp","moduleVersion":1,"operation":"myOperation","params":{"someString":"this very interesting text will be output in uppercase"}}'
```

## Documentation
- [Webhook Instructions](doc/01-webhook.md)
- [Push Instructions](doc/02-push.md)


## Change log
Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.


## Testing
``` bash
$ composer test
```


## About

### Security, bugs and feature requests
Bugs and feature request are tracked on [GitHub](https://github.com/kinobi/n4b/issues)

If you discover any security related issues, please email lbrianto@be-bound.com instead of using the issue tracker.


### Author
Lionel Brianto - <lbrianto@be-bound.com> - <http://twitter.com/kinobiweb>
<!--See also the list of [contributors](https://github.com/kinobi/n4b/contributors) which participated in this project.-->


### License
N4B PHP is licensed under the MIT License - see the `LICENSE` file for details


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [1.1.3] - 2017-02-23
### Changed
- Travis CI config

## [1.1.2] - 2017-02-23
### Changed
- N4B\Webhook Refactoring
- N4B\Push Refactoring

### Fixed
- Fix misleading typos in documentation.

## [1.1.1] - 2016-09-25
### Changed
- N4B\Webhook Log levels tweaked

## [1.1.0] - 2016-09-25
### Added
- Adding PSR-3 Log capabilities

## [1.0.3] - 2016-09-24
### Added
- Webhook run option to get the response as a string instead of sending it

### Changed
- N4B\Error simplified

## [1.0.2] - 2016-09-23
### Added
- This CHANGELOG file.
- Webhook documentation.
- Push documentation.

### Changed
- README updated.

## [1.0.1] - 2016-09-20
### Fixed
- Fix default Webhook run options array.

## 1.0.0 - 2016-09-19
### Added
- First version of the N4B PHP Library.

[Unreleased]: https://github.com/kinobi/n4b/compare/1.1.3...HEAD
[1.1.3]: https://github.com/kinobi/n4b/compare/1.1.2...1.1.3
[1.1.2]: https://github.com/kinobi/n4b/compare/1.1.1...1.1.2
[1.1.1]: https://github.com/kinobi/n4b/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/kinobi/n4b/compare/1.0.3...1.1.0
[1.0.3]: https://github.com/kinobi/n4b/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/kinobi/n4b/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/kinobi/n4b/compare/1.0.0...1.0.1


tests/N4B/PushTest.php
<?php

namespace N4B;

class PushTest extends \PHPUnit_Framework_TestCase
{
    private $beAppName = 'testPush';
    private $beAppId = 13;
    private $beAppVersion = 2;
    private $beAppSecret = 'Sup3rS3cr3tch41n4Push';

    public function testItCanBeInstantiate()
    {
        $n4bPush = $this->instantiatePush();

        $this->assertInstanceOf(Push::class, $n4bPush);
        $this->assertEquals($this->beAppName, $n4bPush->getBeappName());
        $this->assertEquals($this->beAppId, $n4bPush->getBeappId());
        $this->assertEquals($this->beAppVersion, $n4bPush->getBeappVersion());
        $this->assertEquals($this->beAppSecret, $n4bPush->getBeappSecret());
    }

    public function testItCheckTheUrgencyLevel()
    {
        $this->expectException(\OutOfRangeException::class);
        $n4bPush = $this->instantiatePush();
        $n4bPush->send('someNotification', 'userId1', [], rand(100, 200));
    }

    public function testItCanSendPushMessage()
    {
        $n4bPush = $this->instantiatePush();
        $url = sprintf('https://web.be-bound.com/n4bp/%s_%s/%s', $this->beAppName, $this->beAppId, $this->beAppVersion);
        file_put_contents($url, json_encode(['userId1' => 1]));

        $response = $n4bPush->send('someNotification', 'userId1', [], Push::N4B_PUSH_URGENCY_BEBOUND_ONLY);

        $this->assertArrayHasKey('userId1', $response);
    }

    protected function setUp()
    {
        stream_wrapper_unregister("https");
        stream_wrapper_register("https", "N4B\\Mocks\\MockPhpStream");
    }

    protected function tearDown()
    {
        stream_wrapper_restore("https");
    }

    /**
     * @return Push
     */
    private function instantiatePush()
    {
        $n4bPush = new Push($this->beAppName, $this->beAppId, $this->beAppVersion, $this->beAppSecret);

        return $n4bPush;
    }
}


tests/N4B/WebhookTest.php
<?php

namespace N4B;

class WebhookTest extends \PHPUnit_Framework_TestCase
{
    private $beAppName = 'testApp';
    private $beAppId = 13;
    private $beAppVersion = 1;
    private $beAppSecret = 'Sup3rS3cr3tch41n';

    public function testItCanBeInstantiate()
    {
        $n4b = $this->instantiateWebhook();

        $this->assertInstanceOf(Webhook::class, $n4b);
        $this->assertEquals($this->beAppName, $n4b->getBeappName());
        $this->assertEquals($this->beAppId, $n4b->getBeappId());
        $this->assertEquals($this->beAppVersion, $n4b->getBeappVersion());
        $this->assertEquals($this->beAppSecret, $n4b->getBeappSecret());
    }

    public function testItCanBeUpdated()
    {
        $n4b = $this->instantiateWebhook();
        $newBeappName = 'testAppUpdated';
        $newBeappId = 42;
        $newBeappVersion = 2;
        $newBeappSecret = '4n0th3r$3cr3T';

        $n4b->setBeappName($newBeappName);
        $this->assertEquals($newBeappName, $n4b->getBeappName());

        $n4b->setBeappId($newBeappId);
        $this->assertEquals($newBeappId, $n4b->getBeappId());

        $n4b->setBeappVersion($newBeappVersion);
        $this->assertEquals($newBeappVersion, $n4b->getBeappVersion());

        $n4b->setBeappSecret($newBeappSecret);
        $this->assertEquals($newBeappSecret, $n4b->getBeappSecret());
    }

    public function testItCanReturnErrorBB_ERROR_METHOD_NOT_FOUND()
    {
        $n4b = $this->instantiateWebhook();
        $this->setRequestData();

        $response = $n4b->run(['getResponse' => true]);
        $this->assertJsonStringEqualsJsonString('{"error":"BB_ERROR_METHOD_NOT_FOUND"}', $response);
    }

    public function testItCanCatchForgottenExceptions()
    {
        $operationName = 'opWithException';
        $n4b = $this->instantiateWebhook();
        $n4b->add($operationName, function () {
            throw new \RuntimeException('Forget to catch this one');
        });

        $this->setRequestData($operationName);
        $response = $n4b->run(['getResponse' => true]);
        $this->assertJsonStringEqualsJsonString('{"error":"BB_ERROR_UNKNOWN_USER_SPECIFIED_ERROR"}', $response);
    }

    public function testCatchAllOptionCanBeTurnedOff()
    {
        $this->expectException(\RuntimeException::class);
        $operationName = 'opWithException';
        $n4b = $this->instantiateWebhook();
        $n4b->add($operationName, function () {
            throw new \RuntimeException('Forget to catch this one');
        });

        $this->setRequestData($operationName);
        $n4b->run(['catchAll' => false]);
    }

    public function testItcanCheckWrongBasicAuthorisation()
    {
        $n4b = $this->instantiateWebhook();
        $n4b->setBeappSecret('changedSecret');
        $response = $n4b->run(['getResponse' => true]);
        $this->assertJsonStringEqualsJsonString('{"error":"BB_ERROR_AUTHORIZATION"}', $response);
    }

    /**
     * @runInSeparateProcess
     */
    public function testItCanSendResponse()
    {
        $this->expectOutputString('{"params":{"yodaShout":"YOU MUST UNLEARN WHAT YOU HAVE LEARNED."}}');
        $operation = 'shout';
        $n4b = $this->instantiateWebhook();
        $this->setRequestData($operation, ['yodaTalk' => 'you must unlearn what you have learned.']);

        $n4b->add($operation, function ($params) {
            return ['yodaShout' => strtoupper($params['yodaTalk'])];
        });

        $n4b->run(['getResponse' => false]);
    }

    public function testItCanRejectEmptyRequest()
    {
        $n4b = $this->instantiateWebhook();
        file_put_contents('php://input', 'notAJsonString');
        $response = $n4b->run(['getResponse' => true]);
        $this->assertJsonStringEqualsJsonString('{"error":"BB_ERROR_REQUEST_REJECTED"}', $response);
    }

    public function testItCanIgnoreIrrelevantRequest()
    {
        $n4b = $this->instantiateWebhook();
        $this->setRequestData('anyOperation', ['luke' => 'jedi'], 504, 'empireStrikesBack', 5);
        $return = $n4b->run();
        $this->assertNull($return);
    }

    protected function setUp()
    {
        stream_wrapper_unregister("php");
        stream_wrapper_register("php", "N4B\\Mocks\\MockPhpStream");
    }

    protected function tearDown()
    {
        stream_wrapper_restore("php");
    }

    /**
     * @return Webhook
     */
    private function instantiateWebhook()
    {
        $n4b = new Webhook($this->beAppName, $this->beAppId, $this->beAppVersion, $this->beAppSecret);

        return $n4b;
    }

    private function setRequestData(
        $operationName = 'myOperation',
        array $params = [],
        $moduleId = null,
        $moduleName = null,
        $moduleVersion = null
    ) {
        $moduleId = $moduleId ?: $this->beAppId;
        $moduleName = $moduleName ?: $this->beAppName;
        $moduleVersion = $moduleVersion ?: $this->beAppVersion;
        $data = sprintf('{"transport":"web","userId":"id","moduleId":%s,"moduleName":"%s","moduleVersion":%s,"operation":"%s","params":%s}',
            $moduleId, $moduleName, $moduleVersion, $operationName, json_encode($params));
        file_put_contents('php://input', $data);
    }
}


tests/N4B/WebhookHandlerTest.php
<?php

namespace N4B;

class WebhookHandlerTest extends \PHPUnit_Framework_TestCase
{
    private $beAppName = 'testApp';
    private $beAppId = 13;
    private $beAppVersion = 1;
    private $beAppSecret = 'Sup3rS3cr3tch41n';

    public function testItCanBeInstantiate()
    {
        $n4b = new WebhookHandler($this->beAppName, $this->beAppId, $this->beAppVersion, $this->beAppSecret);
        $this->assertInstanceOf(WebhookHandler::class, $n4b);
    }
}


src/N4B/Push.php
<?php

namespace N4B;

class Push extends HandlerAbstract
{
    const N4B_URL_PATTERN = 'https://web.be-bound.com/n4bp/%s_%s/%s';
    const N4B_PUSH_URGENCY_BEBOUND_ONLY = 1;
    const N4B_PUSH_URGENCY_HIGH = 2;
    const N4B_PUSH_URGENCY_DATA_ONLY = 5;

    protected static $availableUrgencies = [
        self::N4B_PUSH_URGENCY_BEBOUND_ONLY,
        self::N4B_PUSH_URGENCY_HIGH,
        self::N4B_PUSH_URGENCY_DATA_ONLY,
    ];

    /**
     * Send a push message through N4B.
     *
     * @param $operation
     * @param string|array $users
     * @param array        $params
     * @param int          $urgency
     *
     * @return array
     */
    public function send($operation, $users, array $params = [], $urgency = self::N4B_PUSH_URGENCY_BEBOUND_ONLY)
    {
        if (!is_array($users)) {
            $users = (array) $users;
        }

        if (!in_array($urgency, static::$availableUrgencies)) {
            $error = 'Push urgency not available';
            $this->logger->critical($error, ['level' => $urgency]);
            throw new \OutOfRangeException($error);
        }

        $context = $this->createContext($operation, $users, $params, $urgency);
        $url = $this->createUrl();
        $response = file_get_contents($url, false, $context);
        $this->logger->info('Push message sent');

        return json_decode($response, true);
    }

    private function createContext($operation, array $users, array $params, $urgency)
    {
        $data = json_encode(
            [
                'operation' => $operation,
                'userId'    => $users,
                'params'    => $params,
                'urgency'   => $urgency,
            ]
        );

        $opts = [
            'http' => [
                'method' => 'POST',
                'header' => [
                    'Content-type: application/json',
                    'Authorization: Basic '.$this->generateBearer(),
                ],
                'content' => $data,
            ],
        ];

        return stream_context_create($opts);
    }

    private function generateBearer()
    {
        $bearer = sprintf('%s_%s:%s', $this->beappName, $this->beappId, $this->beappSecret);

        return base64_encode($bearer);
    }

    private function createUrl()
    {
        return sprintf(static::N4B_URL_PATTERN, $this->beappName, $this->beappId, $this->beappVersion);
    }
}


src/N4B/PushHandler.php
<?php

namespace N4B;

class PushHandler extends HandlerAbstract
{
    const N4B_URL_PATTERN = 'https://web.be-bound.com/n4bp/%s_%s/%s';
    const N4B_PUSH_URGENCY_BEBOUND_ONLY = 1;
    const N4B_PUSH_URGENCY_HIGH = 2;
    const N4B_PUSH_URGENCY_DATA_ONLY = 5;

    protected static $n4bAvailableUrgencies = [
        self::N4B_PUSH_URGENCY_BEBOUND_ONLY,
        self::N4B_PUSH_URGENCY_HIGH,
        self::N4B_PUSH_URGENCY_DATA_ONLY,
    ];

    /**
     * Send a push message through N4B.
     *
     * @param $operation
     * @param string|array $users
     * @param array $params
     * @param int $urgency
     *
     * @return array
     */
    public function send($operation, $users, array $params = [], int $urgency = self::N4B_PUSH_URGENCY_BEBOUND_ONLY)
    {
        if (!is_array($users)) {
            $users = (array)$users;
        }

        if (!in_array($urgency, static::$n4bAvailableUrgencies)) {
            throw new \OutOfRangeException('Push urgency not available');
        }

        $context = $this->createContext($operation, $users, $params, $urgency);
        $url = $this->createUrl();
        $response = file_get_contents($url, false, $context);

        return json_decode($response, true);
    }

    private function createContext($operation, array $users, array $params, int $urgency)
    {
        $data = json_encode(
            [
                'operation' => $operation,
                'userId' => $users,
                'params' => $params,
                'urgency' => $urgency,
            ]
        );

        $opts = [
            'http' =>
                [
                    'method' => 'POST',
                    'header' => [
                        'Content-type: application/json',
                        'Authorization: Basic ' . $this->generateBearer()
                    ],
                    'content' => $data
                ]
        ];

        return stream_context_create($opts);
    }

    private function generateBearer()
    {
        $bearer = sprintf('%s_%s:%s', $this->moduleName, $this->moduleId, $this->modulePassword);

        return base64_encode($bearer);
    }

    private function createUrl()
    {
        return sprintf(static::N4B_URL_PATTERN, $this->moduleName, $this->moduleId, $this->moduleVersion);
    }
}


src/N4B/WebhookHandler.php
<?php

namespace N4B;

use Throwable;

class WebhookHandler extends HandlerAbstract
{
    protected $operationsMap = [];

    /**
     * Handle a N4B request.
     *
     * @param array $options
     *
     * @throws Throwable
     *
     * @internal param bool $authCheck
     * @internal param bool $debug
     */
    public function run(array $options)
    {
        $options = array_merge([
            'authCheck' => true,
            'debug'     => false,
        ], $options);

        try {
            $data = $this->parseRequestBody();

            if ($this->moduleName != $data['moduleName']
                || $this->moduleId != $data['moduleId']
                || $this->moduleVersion != $data['moduleVersion']
            ) {
                return; // This request doesn't concern this handler
            }

            if ((bool) $options['authCheck']) {
                $this->checkAuth();
            }

            if (!array_key_exists($data['operation'], $this->operationsMap)) {
                throw new \Exception('BB_ERROR_METHOD_NOT_FOUND');
            }

            $out = $this->operationsMap[$data['operation']]((array) $data['params'], $data['transport'],
                $data['userId']);
            $this->sendResponse(['params' => $out]);
        } catch (Error $e) {
            $this->sendResponse(['error' => $e->getMessage()]);
        } catch (Throwable $e) {
            if (!(bool) $options['debug']) {
                $this->sendResponse(['error' => 'BB_ERROR_UNKNOWN_USER_SPECIFIED_ERROR']);
            } else {
                throw $e;
            }
        }
    }

    /**
     * Map a callable to an operation describes in the BeApp Manifest.
     *
     * @param $operation
     * @param callable $handler
     */
    public function add($operation, callable $handler)
    {
        $this->operationsMap[$operation] = $handler;
    }

    private function checkAuth()
    {
        $auth_usr = $_SERVER ['PHP_AUTH_USER'] ?? null;
        $auth_pwd = $_SERVER ['PHP_AUTH_PW'] ?? null;
        if ($auth_usr != sprintf('%s_%s', $this->moduleName,
                $this->moduleId) || $auth_pwd != $this->modulePassword
        ) {
            throw new Error('BB_ERROR_AUTHORIZATION');
        }

        return true;
    }

    private function parseRequestBody()
    {
        $data = json_decode(file_get_contents('php://input'), true);
        if (!$data) {
            throw new Error('BB_ERROR_REQUEST_REJECTED');
        }

        return $data;
    }

    private function sendResponse($out)
    {
        $response = json_encode($out);
        header('Content-Type: application/json');
        header('Cache-Control: no-cache, must-revalidate'); // No Cache: HTTP/1.1
        header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // No Cache: in the past
        header('Content-Length: '.strlen($response));
        echo $response;
    }
}


src/N4B/HandlerAbstract.php
<?php

namespace N4B;

use Psr\Log\LoggerAwareInterface;
use Psr\Log\LoggerInterface;
use Psr\Log\NullLogger;

abstract class HandlerAbstract implements LoggerAwareInterface
{
    protected $beappName;
    protected $beappId;
    protected $beappVersion;
    protected $beappSecret;

    /**
     * @var LoggerInterface
     */
    protected $logger;

    public function __construct($beappName, $beappId, $beappVersion, $beappSecret)
    {
        $this->setBeappName($beappName);
        $this->setBeappId($beappId);
        $this->setBeappVersion($beappVersion);
        $this->setBeappSecret($beappSecret);
        $this->setLogger(new NullLogger());
    }

    /**
     * Get the BeApp name.
     *
     * @return string
     */
    public function getBeappName()
    {
        return $this->beappName;
    }

    /**
     * Set the BeApp name.
     *
     * @param string $beappName
     */
    public function setBeappName($beappName)
    {
        $this->beappName = $beappName;
    }

    /**
     * Get the BeApp Id.
     *
     * @return int
     */
    public function getBeappId()
    {
        return $this->beappId;
    }

    /**
     * Set the BeApp Id number.
     *
     * @param mixed $beappId
     */
    public function setBeappId($beappId)
    {
        $this->beappId = intval($beappId);
    }

    /**
     * Get the BeApp version number.
     *
     * @return int
     */
    public function getBeappVersion()
    {
        return $this->beappVersion;
    }

    /**
     * Set the BeApp version number.
     *
     * @param int $beappVersion
     */
    public function setBeappVersion($beappVersion)
    {
        $this->beappVersion = intval($beappVersion);
    }

    /**
     * Get the module auth secret.
     *
     * @return string
     */
    public function getBeappSecret()
    {
        return $this->beappSecret;
    }

    /**
     * Set the module auth secret.
     *
     * @param string $beappSecret
     */
    public function setBeappSecret($beappSecret)
    {
        $this->beappSecret = $beappSecret;
    }

    /**
     * @param LoggerInterface $logger
     *
     * @return null|void
     */
    public function setLogger(LoggerInterface $logger)
    {
        $this->logger = $logger;
    }
}


src/N4B/Webhook.php
<?php

namespace N4B;

use Exception;

class Webhook extends HandlerAbstract
{
    protected $operationsMap = [];

    /**
     * Handle a N4B request.
     *
     * @param array $options
     *
     * @return bool|string
     * @throws Exception
     *
     * @internal param bool $authCheck
     * @internal param bool $debug
     */
    public function run(array $options = [])
    {
        $options = array_merge([
            'authCheck' => true,
            'catchAll' => true,
            'getResponse' => false,
        ], $options);

        try {
            $data = $this->parseRequestBody();

            if ($this->beappName != $data['moduleName']
                || $this->beappId != $data['moduleId']
                || $this->beappVersion != $data['moduleVersion']
            ) {
                $this->logger->notice('The request is not relevant for this webhook.');
                return null; // This request doesn't concern this handler
            }

            $this->logger->info(sprintf('Received operation %s', $data['operation']),
                ['params' => $data['params'], 'transport' => $data['transport'], 'userId' => $data['userId']]);

            $this->checkAuth((bool) $options['authCheck']);

            if (!array_key_exists($data['operation'], $this->operationsMap)) {
                $this->logger->error('No callable mapped to this operation', ['operationName' => $data['operation']]);
                throw new Error('BB_ERROR_METHOD_NOT_FOUND');
            }

            $out = $this->operationsMap[$data['operation']]((array) $data['params'], $data['transport'],
                $data['userId']);
            $n4bResponse = ['params' => $out];
        } catch (Error $e) {
            $n4bResponse = ['error' => $e->getMessage()];
            $this->logger->notice('Returning Be-App Error', $n4bResponse);
        } catch (Exception $e) {
            if (!(bool) $options['catchAll']) {
                throw $e;
            }

            $this->logger->critical('Uncaught exception in the operation handler',
                ['exception', $e, 'operation' => $data['operation']]);
            $n4bResponse = ['error' => 'BB_ERROR_UNKNOWN_USER_SPECIFIED_ERROR'];
        }

        return (bool) $options['getResponse'] ? $this->getResponse($n4bResponse) : $this->sendResponse($n4bResponse);
    }

    /**
     * Map a callable to an operation describes in the BeApp Manifest.
     *
     * @param          $operation
     * @param callable $handler
     */
    public function add($operation, callable $handler)
    {
        $this->logger->debug('Operation handler added', ['operation' => $operation]);
        $this->operationsMap[$operation] = $handler;
    }

    private function checkAuth($checkFlag)
    {
        if (!$checkFlag) {
            return true;
        }

        $authUser = isset($_SERVER ['PHP_AUTH_USER']) ? filter_var($_SERVER ['PHP_AUTH_USER']) : null;
        $authPassword = isset($_SERVER ['PHP_AUTH_PW']) ? filter_var($_SERVER ['PHP_AUTH_PW']) : null;
        if ($authUser != sprintf('%s_%s', $this->beappName,
                $this->beappId) || $authPassword != $this->beappSecret
        ) {
            $this->logger->error('Authentication of the request failed.');
            throw new Error('BB_ERROR_AUTHORIZATION');
        }

        return true;
    }

    private function parseRequestBody()
    {
        $data = json_decode(file_get_contents('php://input'), true);
        if (!$data) {
            $this->logger->error('Empty request body.');
            throw new Error('BB_ERROR_REQUEST_REJECTED');
        }

        return $data;
    }

    private function sendResponse($out)
    {
        $response = json_encode($out);

        $this->logger->info('Returns json HTTP response.');
        header('Content-Type: application/json');
        header('Cache-Control: no-cache, must-revalidate'); // No Cache: HTTP/1.1
        header('Expires: Sat, 26 Jul 1997 05:00:00 GMT'); // No Cache: in the past
        header('Content-Length: ' . strlen($response));
        echo $response;
        return true;
    }

    private function getResponse($out)
    {
        $this->logger->info('Returns string response.');
        return json_encode($out);
    }
}
